### PR TITLE
feat: 🎸 Add initial eslint config

### PR DESCRIPTION
**Pull request by maael**
# Outline / Read this first

Rather than re-reviewing our previous rules, and all new rules, it would be good to migrate to eslint with a config as close as possible to what we currently have, and then start opting in to new rules as we want to. This PR tries to create an eslint config that's as close as possible to our previous tslint one. Any differences are highlighted below, with alternatives where possible.

# Outcome

Done using the `npx tslint-to-eslint-config` utility tool and migration outlined [here](https://code.visualstudio.com/api/advanced-topics/tslint-eslint-migration).

Most of the rules should be intact and behave the same, the differences are outlined below, from the log output by the `tslint-to-eslint-config` tool.

9 ESLint rules behave differently from their TSLint counterparts:
  * prefer-arrow/prefer-arrow-functions:
    - ESLint does not support allowing standalone function declarations.
    - ESLint does not support allowing named functions defined with the function keyword.
  * no-console:
    - Custom console methods, if they exist, will no longer be allowed.
  * no-invalid-this:
    - Functions in methods will no longer be ignored.
  * @typescript-eslint/no-unused-expressions:
    - The TSLint optional config "allow-new" is the default ESLint behavior and will no longer be ignored.
  * no-void:
    - ESLint does not support ignoring arrow function shorthand.
  * eqeqeq:
    - Option "smart" allows for comparing two literal values, evaluating the value of typeof and null comparisons.
  * one-var:
    - Variables declared in for loops will no longer be checked.
  * camelcase:
    - Leading and trailing underscores (_) in variable names will now be ignored.
  * no-underscore-dangle:
    - Leading or trailing underscores (_) on identifiers will now be forbidden.

6 rules are not known by tslint-to-eslint-config to have ESLint equivalents:
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "encoding".
    - 🚫  ⠀No equivalent that I can find.
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-mergeable-namespace".
    - ❗❓⠀**Alternative:** Namespaces are outdated anyway, we could replace this with the [no-namespace](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-namespace.md) rule, disallowing namespaces/custom modules for anything besides describing external apis.
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "prefer-while".
    - 🚫 ⠀No equivalent that I can find.
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "return-undefined".
    - 🚫  ⠀No equivalent that I can find.
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "strict-type-predicates".
    - ❗❓⠀**Alternative:** We could use the [no-unnecessary-condition rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unnecessary-condition.md) to partly replace what this was doing.
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "switch-final-break".
    - 🚫 ⠀No equivalent that I can find.
  * tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "no-parameter-reassignment".
    - 🚫 ⠀No equivalent that I can find.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+     */
+    'no-void': 'error',
+    /**
+     * Enforce variables to be declared either separately in functions
+     */
+    'one-var': ['error', 'never'],
+    /**
+     * Prefer arrow functions, allowing named functions at top level
+     */
+    'prefer-arrow/prefer-arrow-functions': ['error', {allowStandaloneDeclarations: true}],
```
Comment (ForbesLindesay) [inline .eslintrc.js:359]: I think `prefer-arrow-callback` may be a better match for the behaviour we had before/wanted: https://eslint.org/docs/rules/prefer-arrow-callback

We could then also pass in `{allowNamedFunctions: true}`

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+     */
+    '@typescript-eslint/unbound-method': 'error',
+    /**
+     * Warns for any two overloads that could be unified into one by using a union or an optional/rest parameter
+     */
+    '@typescript-eslint/unified-signatures': 'error',
+    /**
+     * Enforce camelCase naming convention
+     */
+    camelcase: 'error',
```
Comment (ForbesLindesay) [inline .eslintrc.js:162]: I don't remember us having this as an enforced thing before?

File: .eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+     */
+    'no-template-curly-in-string': 'error',
+    /**
+     * Disallow Initializing to undefined
+     */
+    'no-undef-init': 'error',
+    /**
+     * Disallow dangling underscores in identifiers
+     */
+    'no-underscore-dangle': 'error',
```
Comment (ForbesLindesay) [inline .eslintrc.js:339]: I think this is new. By default, `allowAfterThis` is `false`, which prevents the "prefix private properties with `_`" convention that we do generally follow.

File: .eslintrc.js (hunk 2 above)
Comment (maael) [reply .eslintrc.js:162]: We have `variable-name` enabled [here](https://github.com/ThreadsStyling/tsconfig/blob/master/tslint.js#L360) which [by default only allows only lowerCamelCased or UPPER_CASED variable names](https://palantir.github.io/tslint/rules/variable-name/).

I'm guessing that's why the tool put in this rule in the conversion.

File: .eslintrc.js (hunk 3 above)
Comment (maael) [reply .eslintrc.js:339]: This has come in [due to our use of the `variable-name` rule](https://github.com/typescript-eslint/tslint-to-eslint-config/blob/bdcc6bcca60f0dbed83c43e54df996bb40dbc7d8/src/rules/converters/variable-name.ts). As I understand it, we don't use `check-format` with it, so in theory, the underscore dangles shouldn't have been allowed?

**Comment by ForbesLindesay** [comment]
I think the fact that we weren't using `check-format` means that both the underscore dangle and camel case rules should be removed. All we were actually doing was banning keywords.

**Comment by jasoniangreen** [comment]
> I think the fact that we weren't using `check-format` means that both the underscore dangle and camel case rules should be removed. All we were actually doing was banning keywords.

That makes sense, I am sure I've seen dangling underscores in our variable names on some rare occasions.

**Comment by maael** [comment]
@ForbesLindesay @jasoniangreen Not sure I agree, by not using `check-format`, we were implicitly opting into the `camcelcase` and `no-underscore-dangle` rules. If we take them out, we make it more permissive than what we had. To stay as close to the rules we had previously, we should include them.

I would lean towards leaving them in, and if they become problematic we can take them out.

**Comment by jasoniangreen** [comment]
Oh I see, you're saying they were the default. Well whatever is closest to what we have I agree with.

**Comment by maael** [comment]
After talking to @diogofcunha it seems that _isn't the case_, but also they seem like reasonable things to include. They are probably added due to a bug in the conversion tool, that didn't properly account for `variable-name` how we have it configured. I will leave it as is though, and if we find it problematic we can take it out.

I've included `no-unnecessary-condition` to replace `strict-type-predicates`, and also I've replaced the `prefer-arrow/prefer-arrow-functions` with `prefer-arrow-callback` (including the `{allowNamedFunctions: true}` option).

This should give us a good place to start, and any issues that arise can be fixed.

I'll merge this now.

**Pull request merged**